jetty-http2/http2-http-client-transport/src/test/java/org/eclipse/jetty/http2/client/http/MultiplexedConnectionPoolTest.java

```diff
@@ -85,14 +89,127 @@ public void disposeClient() throws Exception
+                int req = Integer.parseInt(target.substring(1));
+                try
+                {
+                    LOG.debug("req {} is executing", req);
+                    reqExecutingLatches[req].countDown();
+                    Thread.sleep(250);
+                    reqExecutedLatches[req].countDown();
+                    LOG.debug("req {} executed", req);
+
+                    assertTrue(reqFinishingLatches[req].await(5, TimeUnit.SECONDS));
+
+                    response.getWriter().println("req " + req + " executed");
+                    response.getWriter().flush();
+                    LOG.debug("req {} successful", req);
+                }
+                catch (Exception e)
+                {
+                    throw new ServletException(e);
+                }
+            }
+        });
+
+        ClientConnector clientConnector = new ClientConnector();
+        HttpClientTransport transport = new HttpClientTransportOverHTTP2(new HTTP2Client(clientConnector));
+        transport.setConnectionPoolFactory(factory.factory);
+        client = new HttpClient(transport);
+        client.start();
+
+        CountDownLatch[] reqClientDoneLatches = new CountDownLatch[] {new CountDownLatch(1), new CountDownLatch(1), new CountDownLatch(1)};
+
+        sendRequest(reqClientDoneLatches, 0);
+        // wait until handler is executing
+        assertTrue(reqExecutingLatches[0].await(5, TimeUnit.SECONDS));
+        LOG.debug("req 0 executing");
+
+        sendRequest(reqClientDoneLatches, 1);
+        // wait until handler executed sleep
+        assertTrue(reqExecutedLatches[1].await(5, TimeUnit.SECONDS));
+        LOG.debug("req 1 executed");
+
+        // Now the pool contains one connection that is expired but in use by 2 threads.
+
+        sendRequest(reqClientDoneLatches, 2);
+        LOG.debug("req2 sent");
+        assertTrue(reqExecutingLatches[2].await(5, TimeUnit.SECONDS));
+        LOG.debug("req2 executing");
+
+        // The 3rd request has tried the expired request and marked it as closed as it has expired, then used a 2nd one.
+
+        // release and wait for req2 to be done before releasing req1
+        reqFinishingLatches[2].countDown();
+        assertTrue(reqClientDoneLatches[2].await(5, TimeUnit.SECONDS));
+        reqFinishingLatches[1].countDown();
+
+        // release req0 once req1 is done; req 1 should not have closed the response as req 0 is still running
+        assertTrue(reqClientDoneLatches[1].await(5, TimeUnit.SECONDS));
+        reqFinishingLatches[0].countDown();
+        assertTrue(reqClientDoneLatches[0].await(5, TimeUnit.SECONDS));
+
+        // Check that the pool created 2 and removed 2 connections;
+        // 2 were removed b/c waiting for req 2 means the 2nd connection
+        // expired and has to be removed and closed upon being returned to the pool.
+        assertThat(poolCreateCounter.get(), Matchers.is(2));
+        assertThat(poolRemoveCounter.get(), Matchers.is(2));
```

Comment:
Please check also that the pool has 0 connections.

Reply:
Done.